src/Explorer/Controls/SmartUi/SmartUiComponent.tsx

```diff
@@ -101,42 +116,167 @@ export class SmartUiComponent extends React.Component<SmartUiComponentProps, Sma
     fontSize: 12
   };
 
+  componentDidMount(): void {
+    this.setDefaultValues();
+  }
+
   constructor(props: SmartUiComponentProps) {
     super(props);
     this.state = {
+      baselineValues: new Map(),
       currentValues: new Map(),
-      errors: new Map()
+      errors: new Map(),
+      isRefreshing: false
     };
   }
 
+  private setDefaultValues = async (): Promise<void> => {
+    this.setState({ isRefreshing: true });
+    await this.setDefaults(this.props.descriptor.root);
+    this.setState({ isRefreshing: false });
+    await this.initialize();
+  };
+
+  private initialize = async (): Promise<void> => {
+    this.setState({ isRefreshing: true });
+    let { currentValues, baselineValues } = this.state;
+    const initialValues = await this.props.descriptor.initialize();
+    for (const key of initialValues.keys()) {
+      if (this.props.descriptor.inputNames.indexOf(key) === -1) {
+        console.log(this.props.descriptor.inputNames);
+        this.setState({ isRefreshing: false });
+        throw new Error(`${key} is not an input property of this class.`);
+      }
+      currentValues = currentValues.set(key, initialValues.get(key));
+      baselineValues = baselineValues.set(key, initialValues.get(key));
+    }
+    this.setState({ currentValues: currentValues, baselineValues: baselineValues, isRefreshing: false });
+  };
+
+  private discard = (): void => {
+    let { currentValues } = this.state;
+    const { baselineValues } = this.state;
+    for (const key of baselineValues.keys()) {
+      currentValues = currentValues.set(key, baselineValues.get(key));
+    }
+    this.setState({ currentValues: currentValues });
+  };
+
+  private setDefaults = async (currentNode: Node): Promise<void> => {
+    if (currentNode.info && currentNode.info instanceof Function) {
+      currentNode.info = await (currentNode.info as infoPromise)();
+    }
+
+    if (currentNode.input) {
+      currentNode.input = await this.getModifiedInput(currentNode.input);
+    }
+    const promises = currentNode.children?.map(async (child: Node) => await this.setDefaults(child));
+    if (promises) {
+      await Promise.all(promises);
+    }
+  };
+
+  private getModifiedInput = async (input: AnyInput): Promise<AnyInput> => {
```

Comment:
All of the logic in this function could be greatly simplified if you wrote a generic function that returns a value or calls+awaits the function that produces this value. This concept is called a `Thunk` https://en.wikipedia.org/wiki/Thunk. Roughly:

``` ts
export async function resolveThunk<T>(thunk: T | (() => T)): Promise<T> {
  if (thunk instanceof Function) {
    return thunk();
  }
  return thunk;
}
```